src/common/pf_scheduler.h

```diff
@@ -30,38 +30,122 @@ extern "C" {
+ */
+const char * pf_scheduler_get_name (const pf_scheduler_handle_t * handle);
+
 /**
  * Schedule a call-back at a specific time.
  *
+ * The callback must update the internal state stored in the handle. That is
+ * done by calling \a pf_scheduler_reset_handle() or again calling
+ * \ pf_scheduler_add().
+ *
  * @param net              InOut: The p-net stack instance
  * @param delay            In:    The delay until the function shall be called,
  *                                in microseconds. Max
  *                                PF_SCHEDULER_MAX_DELAY_US.
- * @param p_name           In:    Caller/owner (for removal and debugging).
  * @param cb               In:    The call-back.
  * @param arg              In:    Argument to the call-back.
- * @param p_timeout        Out:   The timeout instance (used to remove if
- *                                necessary).
+ * @param handle           InOut: Timeout handle.
  * @return  0  if the call-back was scheduled.
  *          -1 if an error occurred.
  */
 int pf_scheduler_add (
    pnet_t * net,
    uint32_t delay,
-   const char * p_name,
    pf_scheduler_timeout_ftn_t cb,
    void * arg,
-   uint32_t * p_timeout);
+   pf_scheduler_handle_t * handle);
+
+/**
+ * Re-schedule a call-back at a specific time.
+ *
+ * Removes the already scheduled instance, if existing.
+ *
+ * Do not use in a scheduler callback.
+ *
+ * @param net              InOut: The p-net stack instance
+ * @param delay            In:    The delay until the function shall be called,
+ *                                in microseconds. Max
+ *                                PF_SCHEDULER_MAX_DELAY_US.
+ * @param cb               In:    The call-back.
+ * @param arg              In:    Argument to the call-back.
+ * @param handle           InOut: Timeout handle.
+ * @return  0  if the call-back was scheduled.
+ *          -1 if an error occurred.
+ */
+int pf_scheduler_restart (
+   pnet_t * net,
+   uint32_t delay,
+   pf_scheduler_timeout_ftn_t cb,
+   void * arg,
+   pf_scheduler_handle_t * handle);
+
+/**
+ * Stop a timeout, if it is running.
+ *
+ * Silently ignoring the timeout if not running.
+ *
+ * @param net              InOut: The p-net stack instance
+ * @param handle           InOut: Timeout handle.
+ */
+void pf_scheduler_remove_if_running (
+   pnet_t * net,
+   pf_scheduler_handle_t * handle);
 
 /**
- * Stop a timeout. If it is not scheduled then ignore.
+ * Stop a timeout.
  * @param net              InOut: The p-net stack instance
- * @param p_name           In:    Must be exactly the same address as in the
- *                                pf_scheduler_add().
- * @param timeout          In:    Time instance to remove
- *                                (see pf_scheduler_add)
+ * @param handle           InOut: Timeout handle.
  */
-void pf_scheduler_remove (pnet_t * net, const char * p_name, uint32_t timeout);
+void pf_scheduler_remove (pnet_t * net, pf_scheduler_handle_t * handle);
```

Comment:
If this operation did print a warning instead of an error maybe the pf_scheduler_remove_if_running would not be needed?

Reply:
I think there is room for functions with both behaviors